pkg/package-server/provider/registry.go

```diff
@@ -233,66 +387,52 @@ func (p *RegistryProvider) Get(namespace, name string) (*operators.PackageManife
 }
 
 func (p *RegistryProvider) List(namespace string) (*operators.PackageManifestList, error) {
-	logger := logrus.WithFields(logrus.Fields{
-		"action":    "List PackageManifests",
-		"namespace": namespace,
-	})
-
-	p.mu.RLock()
-	defer p.mu.RUnlock()
-
-	pkgs := []operators.PackageManifest{}
-	for _, client := range p.clients {
-		if client.source.GetNamespace() == namespace || client.source.GetNamespace() == p.globalNamespace || namespace == metav1.NamespaceAll {
-			logger.Debugf("found CatalogSource %s", client.source.GetName())
+	var pkgs []*operators.PackageManifest
+	if namespace == metav1.NamespaceAll {
+		all, err := p.pkgLister.List(labels.Everything())
+		if err != nil {
+			return nil, err
+		}
+		pkgs = append(pkgs, all...)
+	} else {
+		nsPkgs, err := p.pkgLister.PackageManifests(namespace).List(labels.Everything())
+		if err != nil {
+			return nil, err
+		}
+		pkgs = append(pkgs, nsPkgs...)
 
-			stream, err := client.ListPackages(context.Background(), &api.ListPackageRequest{})
+		if namespace != p.globalNamespace {
+			globalPkgs, err := p.pkgLister.PackageManifests(p.globalNamespace).List(labels.Everything())
 			if err != nil {
-				logger.WithField("err", err.Error()).Warnf("error getting stream")
-				continue
-			}
-			for {
-				pkgName, err := stream.Recv()
-				if err == io.EOF {
-					break
-				}
-
-				if err != nil {
-					logger.WithField("err", err.Error()).Warnf("error getting data")
-					break
-				}
-				pkg, err := client.GetPackage(context.Background(), &api.GetPackageRequest{Name: pkgName.GetName()})
-				if err != nil {
-					logger.WithField("err", err.Error()).Warnf("error getting package")
-					continue
-				}
-				newPkg, err := toPackageManifest(logger, pkg, client)
-				if err != nil {
-					logger.WithField("err", err.Error()).Warnf("eliding package: error converting to packagemanifest")
-					continue
-				}
-
-				// Set request namespace to stop kube clients from complaining about global namespace mismatch.
-				if namespace != metav1.NamespaceAll {
-					newPkg.SetNamespace(namespace)
-				}
-				pkgs = append(pkgs, *newPkg)
+				return nil, err
 			}
+
+			pkgs = append(pkgs, globalPkgs...)
+		}
+	}
+
+	pkgList := &operators.PackageManifestList{}
+	for _, pkg := range pkgs {
+		out := pkg.DeepCopy()
+		// Set request namespace to stop k8s clients from complaining about namespace mismatch.
+		if namespace != metav1.NamespaceAll {
+			out.SetNamespace(namespace)
 		}
+		pkgList.Items = append(pkgList.Items, *out)
 	}
 
-	return &operators.PackageManifestList{Items: pkgs}, nil
+	return pkgList, nil
 }
 
-func toPackageManifest(logger *logrus.Entry, pkg *api.Package, client registryClient) (*operators.PackageManifest, error) {
+func newPackageManifest(ctx context.Context, logger *logrus.Entry, pkg *api.Package, client *registryClient) (*operators.PackageManifest, error) {
```

Comment:
Plumbed down the context so that cache building can enforce a timeout on packagemanifest generation.